src/metabase/task/send_pulses.clj

```diff
@@ -1,136 +1,197 @@
+      (log/debugf "Starting Pulse Execution: %d" pulse-id)
+      (when-let [pulse (pulse/retrieve-notification pulse-id :archived false)]
+        (metabase.pulse/send-pulse! pulse :channel-ids channel-ids))
+      (log/debugf "Finished Pulse Execution: %d" pulse-id))
+    (catch Throwable e
+      (log/errorf e "Error sending Pulse %d to channel ids: %s" pulse-id (str/join ", " channel-ids)))))
+
+(mu/defn ^:private send-pulse-trigger
+  "Build a Quartz trigger to send a pulse to a list of channel-ids."
+  ^CronTrigger
+  [pulse-id     :- pos-int?
+   schedule-map :- map?
+   pc-ids       :- [:maybe [:set pos-int?]]]
+  (when (seq pc-ids)
+    (triggers/build
+     (triggers/with-identity (send-pulse-trigger-key pulse-id schedule-map))
+     (triggers/for-job send-pulse-job-key)
+     (triggers/using-job-data {"pulse-id"    pulse-id
+                               "channel-ids" (set pc-ids)})
+     (triggers/with-schedule
+       (cron/schedule
+        (cron/cron-schedule (u.cron/schedule-map->cron-string schedule-map))
+        ;; if we miss a sync for one reason or another (such as system being down) do not try to run the sync again.
+        ;; Just wait until the next sync cycle.
+        ;;
+        ;; See https://www.nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html for more info
+        (cron/with-misfire-handling-instruction-ignore-misfires))))))
+
+(jobs/defjob ^{:doc "Triggers that send a pulse to a list of channels at a specific time"}
+  SendPulse
+  [context]
+  (let [{:strs [pulse-id channel-ids]} (qc/from-job-data context)]
+    (send-pulse! pulse-id channel-ids)))
+
+;;; --------------------------------------------- Job: RePrioritizeSendPulses -------------------------------------------
+
+(def ^:private reprioritize-send-pulse-job-key (jobs/key "metabase.task.send-pulses.reprioritize.job"))
+
+(def ^:private reprioritize-send-pulse-trigger-key (triggers/key "metabase.task.send-pulses.reprioritize.trigger"))
+
+(defn update-send-pulse-trigger-if-needed!
+  "Update send pulse trigger of a pulse for a specific schedule map with new pulse channel ids.
+
+  Send Pulse triggers are grouped by pulse id and schedule time, meaning PulseChannels of a Pulse that scheduled
+  to run at the same time of will be send together.
+  This function will updates the corresponding trigger if PulseChannels changes.
+
+  * To add 2 pulse channels to a trigger
+    (update-send-pulse-trigger-if-needed! pulse-id schedule-map :add-pc-ids #{1 2}))
+
+  * To remove 2 pulse channels from a trigger
+    (update-send-pulse-trigger-if-needed! pulse-id schedule-map :remove-pc-ids #{1 2}))"
+  [pulse-id schedule-map & {:keys [add-pc-ids remove-pc-ids]}]
+  (let [schedule-map     (update-vals schedule-map
+                                      #(if (keyword? %)
+                                         (name %)
+                                         %))
+        trigger-key      (send-pulse-trigger-key pulse-id schedule-map)
```

Comment:
A small thing: to prep for milestone 2, we're going to need to change this function to take a cron-string instead of a schedule-map. So we might as well start with it taking a cron string so we don't have to change it later. In both places this is called we calculate the cron string from the schedule map anyway, so making this take a cron string is less code and will require fewer changes in the future.

Reply:
good idea, but it's actually used in many PulseChannel hooks. Also, `u.cron/schedule-map->cron-string` accepts a map where values are string, but `schedule-map` is keyworded by toucan2 hooks for PulseChannel, so I have to manually turn the value to str. 
```clojure
(update-vals schedule-map
                                      #(if (keyword? %)
                                         (name %)
                                         %))
```

with this I think we'ok with updating this function argument when we get to the 2nd milestone.